pulp_ansible/app/downloaders.py

```diff
@@ -0,0 +1,83 @@
+from logging import getLogger
+import asyncio
+import backoff
+import json
+
+from aiohttp.client_exceptions import ClientResponseError
+
+from pulpcore.plugin.download import http_giveup, HttpDownloader
+
+
+log = getLogger(__name__)
+
+
+class TokenAuthHttpDownloader(HttpDownloader):
+    """
+    Custom Downloader that automatically handles Token Based and Basic Authentication.
+
+    """
+
+    token_lock = asyncio.Lock()
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the downloader.
+        """
+        self.remote = kwargs.pop("remote")
```

Comment:
This is definitly how the current ones implement this, but at a pulpcore we were hoping to have it be done differently in the future. This pattern would be replaced with a subclassed DownloadFactory and the kwargs would become positional args passed here instead. I wrote more about this change here: https://pulp.plan.io/issues/6965#note-2